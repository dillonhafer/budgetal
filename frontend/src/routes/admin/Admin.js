import React, { Component } from 'react';
import { title, notice, error } from 'window';

import {
  AdminUsersRequest,
  AdminTestEmailRequest,
  AdminErrorRequest,
  AdminTestPushNotificationRequest,
} from '@shared/api/admin';
import moment from 'moment';

import {
  Avatar,
  Button,
  Dialog,
  Heading,
  Pane,
  Table,
  TextInputField,
} from 'evergreen-ui';

export default class Admin extends Component {
  state = {
    isAdmin: false,
    testEmailLoading: false,
    errorLoading: false,
    pushNotificationLoading: false,
    pushNotificationVisible: false,
    users: [],
  };

  componentDidMount() {
    this.checkForAdmin();
  }

  testPushNotification = async e => {
    e.preventDefault();
    const titleField = document.getElementsByName('pnTitle')[0];
    const bodyField = document.getElementsByName('pnBody')[0];
    this.setState({ pushNotificationLoading: true });
    try {
      const title = titleField.value;
      const body = bodyField.value;

      const resp = await AdminTestPushNotificationRequest({ title, body });
      if (resp && resp.ok) {
        notice('Sent Push Notification');
        this.setState({
          pushNotificationVisible: false,
        });
        titleField.value = '';
        bodyField.value = '';
      }
    } catch (err) {
      console.log(err);
    } finally {
      this.setState({
        pushNotificationLoading: false,
      });
    }
  };

  sendTestEmail = async () => {
    this.setState({ testEmailLoading: true });
    try {
      const resp = await AdminTestEmailRequest();
      if (resp && resp.ok) {
        notice('Test email has been sent');
      } else {
        error('Could not send test email');
      }
    } catch (err) {
      console.log(err);
    } finally {
      this.setState({ testEmailLoading: false });
    }
  };

  sendErrorTest = async () => {
    this.setState({ errorLoading: true });
    try {
      await AdminErrorRequest();
    } catch (err) {
      console.log(err);
    } finally {
      this.setState({ errorLoading: false });
    }
  };

  checkForAdmin = async () => {
    const resp = await AdminUsersRequest();
    const isAdmin = resp && resp.ok && resp.users.length;
    if (isAdmin) {
      title(`Admin`);
      this.setState({
        isAdmin,
        users: resp.users,
      });
    } else {
      this.props.history.replace('404');
    }
  };

  render() {
    const {
      isAdmin,
      testEmailLoading,
      errorLoading,
      pushNotificationLoading,
    } = this.state;
    if (!isAdmin) {
      return null;
    }

    return (
      <Pane paddingLeft={24} paddingRight={24}>
        <Pane
          display="flex"
          flexDirection="row"
          alignItems="center"
          justifyContent="space-between"
          marginBottom={16}
        >
          <Heading size={800}>ADMIN PANEL</Heading>
        </Pane>
        <Pane
          display="flex"
          padding={16}
          marginBottom={8}
          background="tint2"
          borderRadius={3}
        >
          <Pane flex={1} alignItems="center" display="flex">
            <Heading size={600}>TEST CONFIG</Heading>
          </Pane>
        </Pane>
        <Pane>
          <Button
            height={40}
            iconBefore="envelope"
            appearance="primary"
            intent="none"
            disabled={testEmailLoading}
            onClick={this.sendTestEmail}
          >
            {testEmailLoading ? 'Sending...' : 'Send Test Email'}
          </Button>
          &nbsp;
          <Button
            height={40}
            iconBefore="error"
            appearance="primary"
            intent="danger"
            disabled={errorLoading}
            onClick={this.sendErrorTest}
          >
            {errorLoading ? 'Loading...' : 'Test 500'}
          </Button>
          &nbsp;
          <Button
            height={40}
            iconBefore="notifications"
            appearance="primary"
            intent="none"
            onClick={() => this.setState({ pushNotificationVisible: true })}
          >
            Test Push Notification
          </Button>
          <Dialog
            preventBodyScrolling
            isShown={this.state.pushNotificationVisible}
            title="Test Push Notification"
            width={350}
            onCloseComplete={() => {
              this.setState({ pushNotificationVisible: false });
            }}
            confirmLoading={pushNotificationLoading}
            cancelText="Cancel"
            onConfirm={this.testPushNotification}
            confirmLabel={
              pushNotificationLoading ? 'Loading...' : 'Send Notification'
            }
          >
            <form onSubmit={this.testPushNotification}>
              <TextInputField required label="Title" name="pnTitle" />
              <TextInputField required label="Body" name="pnBody" />
              <input type="submit" className="hide" value="Submit" />
            </form>
          </Dialog>
        </Pane>

        <Pane height="48px" />

        <Pane
          display="flex"
          padding={16}
          marginBottom={8}
          background="tint2"
          borderRadius={3}
        >
          <Pane flex={1} alignItems="center" display="flex">
            <Heading size={600}>USERS</Heading>
          </Pane>
        </Pane>
        <Table>
          <Table.Head accountForScrollbar>
            <Table.TextHeaderCell>Name</Table.TextHeaderCell>
            <Table.TextHeaderCell>Email</Table.TextHeaderCell>
            <Table.TextHeaderCell>Last Sign In</Table.TextHeaderCell>
            <Table.TextHeaderCell>IP</Table.TextHeaderCell>
            <Table.TextHeaderCell>#</Table.TextHeaderCell>
          </Table.Head>
          <Table.Body>
            {this.state.users.map(user => (
              <Table.Row key={`${user.email}`}>
                <Table.TextCell>
                  <Pane display="flex" alignItems="center" flexDirection="row">
                    <Avatar
                      {...(/.*missing-profile.*/.test(user.avatarUrl)
                        ? {}
                        : { src: user.avatarUrl })}
                      name={`${user.firstName || '?'} ${user.lastName || '?'}`}
                      size={40}
                      marginRight={16}
                    />
                    <span>
                      {user.firstName || '-'}, <b>{user.lastName || '-'}</b>
                    </span>
                  </Pane>
                </Table.TextCell>
                <Table.TextCell>{user.email}</Table.TextCell>
                <Table.TextCell>
                  {moment(user.lastSignIn).format('MMMM DD, YYYY - h:mm:ss a')}
                </Table.TextCell>
                <Table.TextCell isNumber>{user.ip}</Table.TextCell>
                <Table.TextCell isNumber>{user.signInCount}</Table.TextCell>
              </Table.Row>
            ))}
          </Table.Body>
        </Table>
      </Pane>
    );
  }
}
