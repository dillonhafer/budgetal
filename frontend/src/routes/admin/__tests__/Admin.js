import React from 'react';
import { shallow } from 'enzyme';
import Admin from '../Admin';

const user = {
  id: 1,
  firstName: 'Liz',
  lastName: 'Lemon',
  email: 'lizzy+lemon@example.com',
  lastSignIn: '2017-11-13T18:20:44.114363-05:00',
};

describe('Admin', () => {
  it('renders', () => {
    const comp = shallow(<Admin />);
    comp.setState({ isAdmin: true, users: [user] });
    expect(comp).toMatchSnapshot();
  });
});
