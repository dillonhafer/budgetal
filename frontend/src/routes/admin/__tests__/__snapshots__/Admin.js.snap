// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Admin renders 1`] = `
<withTheme(Pane)>
  <withTheme(Pane)
    alignItems="center"
    display="flex"
    flexDirection="row"
    justifyContent="space-between"
    marginBottom={16}
  >
    <withTheme(Heading)
      size={800}
    >
      ADMIN PANEL
    </withTheme(Heading)>
  </withTheme(Pane)>
  <withTheme(Pane)
    background="tint2"
    borderRadius={3}
    display="flex"
    marginBottom={8}
    padding={16}
  >
    <withTheme(Pane)
      alignItems="center"
      display="flex"
      flex={1}
    >
      <withTheme(Heading)
        size={600}
      >
        TEST CONFIG
      </withTheme(Heading)>
    </withTheme(Pane)>
  </withTheme(Pane)>
  <withTheme(Pane)>
    <withTheme(Button)
      appearance="primary"
      disabled={false}
      height={40}
      iconBefore="envelope"
      intent="none"
      onClick={[Function]}
    >
      Send Test Email
    </withTheme(Button)>
     
    <withTheme(Button)
      appearance="primary"
      disabled={false}
      height={40}
      iconBefore="error"
      intent="danger"
      onClick={[Function]}
    >
      Test 500
    </withTheme(Button)>
     
    <withTheme(Button)
      appearance="primary"
      height={40}
      iconBefore="notifications"
      intent="none"
      onClick={[Function]}
    >
      Test Push Notification
    </withTheme(Button)>
    <withTheme(Dialog)
      cancelText="Cancel"
      confirmLabel="Send Notification"
      confirmLoading={false}
      isShown={false}
      onCloseComplete={[Function]}
      onConfirm={[Function]}
      preventBodyScrolling={true}
      title="Test Push Notification"
      width={350}
    >
      <form
        onSubmit={[Function]}
      >
        <TextInputField
          inputHeight={32}
          inputWidth="100%"
          label="Title"
          name="pnTitle"
          required={true}
        />
        <TextInputField
          inputHeight={32}
          inputWidth="100%"
          label="Body"
          name="pnBody"
          required={true}
        />
        <input
          className="hide"
          type="submit"
          value="Submit"
        />
      </form>
    </withTheme(Dialog)>
  </withTheme(Pane)>
  <withTheme(Pane)
    height="48px"
  />
  <withTheme(Pane)
    background="tint2"
    borderRadius={3}
    display="flex"
    marginBottom={8}
    padding={16}
  >
    <withTheme(Pane)
      alignItems="center"
      display="flex"
      flex={1}
    >
      <withTheme(Heading)
        size={600}
      >
        USERS
      </withTheme(Heading)>
    </withTheme(Pane)>
  </withTheme(Pane)>
  <Table>
    <TableHead
      accountForScrollbar={true}
      height={32}
    >
      <TextTableHeaderCell>
        Name
      </TextTableHeaderCell>
      <TextTableHeaderCell>
        Email
      </TextTableHeaderCell>
      <TextTableHeaderCell>
        Last Sign In
      </TextTableHeaderCell>
      <TextTableHeaderCell>
        IP
      </TextTableHeaderCell>
      <TextTableHeaderCell>
        #
      </TextTableHeaderCell>
    </TableHead>
    <TableBody>
      <withTheme(TableRow)
        key="lizzy+lemon@example.com"
      >
        <TextTableCell
          isNumber={false}
        >
          <withTheme(Pane)
            alignItems="center"
            display="flex"
            flexDirection="row"
          >
            <withTheme(Avatar)
              marginRight={16}
              name="Liz Lemon"
              size={40}
            />
            <span>
              Liz
              , 
              <b>
                Lemon
              </b>
            </span>
          </withTheme(Pane)>
        </TextTableCell>
        <TextTableCell
          isNumber={false}
        >
          lizzy+lemon@example.com
        </TextTableCell>
        <TextTableCell
          isNumber={false}
        >
          November 13, 2017 - 6:20:44 pm
        </TextTableCell>
        <TextTableCell
          isNumber={true}
        />
        <TextTableCell
          isNumber={true}
        />
      </withTheme(TableRow)>
    </TableBody>
  </Table>
</withTheme(Pane)>
`;
